File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        frequencies = defaultdict(int)
        for num in nums:
            frequencies[num] += 1
        
        pairs, k_freq = ((value, key) for key, value in frequencies.items()), []
        
        for pair in pairs:
            if len(k_freq) < k:
                heappush(k_freq, pair)
            elif len(k_freq) >= k and k_freq[0][0] < pair[0]:
                heappop(k_freq)
                heappush(k_freq, pair)
            
        return [key for val, key in k_freq]

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> frequencies;
        for (int num : nums) ++frequencies[num];
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> heap;
        
        for (auto& [key, val] : frequencies) {
            if (heap.size() < k) heap.push({val, key});
            else if (heap.size() >= k && heap.top().first < val){
                heap.pop();
                heap.push({val, key});
            }
        }
        vector<int> ans;
        while(!heap.empty()) {
            ans.push_back(heap.top().second);
            heap.pop();
        }
        
        return ans;
    }
};